packages/query-core/src/infiniteQueryBehavior.ts

```diff
@@ -53,10 +53,23 @@ export function infiniteQueryBehavior<
           page: unknown,
           previous?: boolean,
         ) => {
-          newPageParams = previous
-            ? [param, ...newPageParams]
-            : [...newPageParams, param]
-          return previous ? [page, ...pages] : [...pages, page]
+          const { maxPages } = context.options
+
+          const limitedPages = new LimitedLengthArray(maxPages, [...pages])
+          const limitedNewPageParams = new LimitedLengthArray(
+            maxPages,
+            newPageParams,
```

Comment:
we're not shallow copying `newPageParams` here 🤔

Reply:
I've tried to stay consistent with the existing code where `newPageParams` value is directly updated.

Reply:
Refactored with the util functions in 434af38abd73ebc4a38a72ca87fbca8f359dbb7a